File: src/main/scala/Race/common/Point.scala
package Race.common

import java.awt

object Point {
  implicit def awt2Point(p: awt.Point): Point = apply(p)
  implicit def point2awt(p: Point): awt.Point = new awt.Point(p.x, p.y)

  def apply(p: awt.Point): Point = PointImp(p.x, p.y)
  def apply(x: Int, y: Int): Point = PointImp(x, y)

  def midPoint(p: (Point, Point)): Point = midPoint(p._1, p._2)
  def midPoint(p1: Point, p2: Point): Point = PointImp((p1.x + p2.x) / 2, (p1.y + p2.y) / 2)

  def distance(p1: Point, p2: Point): Double =
    Math.sqrt(Math.pow(Math.abs(p1.x - p2.x), 2) + Math.pow(Math.abs(p1.y - p2.y), 2))

  def inSquare(p: Point, p1: Point, p2: Point): Boolean = {
    if ((p.x > p1.x) && (p.x > p2.x)) return false
    if ((p.x < p1.x) && (p.x < p2.x)) return false
    if ((p.y > p1.y) && (p.y > p2.y)) return false
    if ((p.y < p1.y) && (p.y < p2.y)) return false
    true
  }

  case class PointImp(x: Int, y: Int) extends Point {
    override def +(other: Point): Point = PointImp(x + other.x, y + other.y)
    override def -(other: Point): Point = PointImp(x - other.x, y - other.y)
    override def rotate(origin: Point, rad: Double): Point = (this - origin).rotate(rad) + this
    override def rotate(rad: Double): Point = {
      PointImp(Math.round((x * Math.cos(rad)) - (y * Math.sin(rad))).toInt, Math.round((y * Math.cos(rad)) + (x * Math.sin(rad))).toInt)
    }

    // from the the x axis
    override def angleFromOrigin(): Double = {
      if (x == 0) return if (y > 0) Math.PI * 0.5 else Math.PI * -0.5

      val angle = Math.atan(y/x)

      if (x > 0) return angle

      if (y < 0) -Math.PI + angle else Math.PI + angle
    }

    override def distanceFromOrigin(): Double = Math.sqrt(Math.pow(x, 2) + Math.pow(y, 2))
  }
}

trait Point {
  def x: Int
  def y: Int

  def +(other: Point): Point
  def -(other: Point): Point
  def rotate(rad: Double): Point
  def rotate(origin: Point, rad: Double): Point

  def angleFromOrigin(): Double
  def distanceFromOrigin(): Double
}

File: PyCar/common/point.py
from PyQt5.QtCore import QPoint
from math import cos, sin, pi, atan, sqrt, pow, fabs


class Point(QPoint):
    def __init__(self, x: int, y: int):
        super().__init__(x, y)

    def __add__(self, other: QPoint):
        return Point(self.x() + other.x(), self.y() + other.y())

    def __sub__(self, other: QPoint):
        return Point(self.x() - other.x(), self.y() - other.y())

    def rotate(self, *args):
        """usage:
        rotate(self, rad) rad = angle in radians
        rotate(self, origin, rad) origin = center of rotation, rad = angle in radians"""
        if len(args) == 1:
            return Point(int(round((self.x() * cos(args[0])) - (self.y() * sin(args[0])))),
                         int(round((self.y() * cos(args[0])) + (self.x() * sin(args[0])))))
        if len(args) == 2 and isinstance(args[0], QPoint):
            return (self - args[0]).rotate(args[1]) + self

    def angle_from_origin(self) -> float:
        """angle from the the x axis"""
        if self.x() == 0:
            if self.y() > 0:
                return pi * 0.5
            else:
                return pi * -0.5

        angle = atan(self.y() / self.x())

        if self.x() > 0:
            return angle

        if self.y() < 0:
            return - pi + angle
        else:
            return pi + angle

    def distance_from_origin(self) -> float:
        return sqrt(pow(self.x(), 2) + pow(self.y(), 2))

    @staticmethod
    def distance(p1: QPoint, p2: QPoint) -> float:
        return sqrt(pow(fabs(p1.x() - p2.x()), 2) + pow(fabs(p1.y() - p2.y()), 2))

    @staticmethod
    def midpoint(p1: QPoint, p2: QPoint):
        return Point((p1.x() + p2.x()) / 2, (p1.y() + p2.y()) / 2)

    @staticmethod
    def in_box(p: QPoint, p1: QPoint, p2: QPoint) -> bool:
        if (p.x() > p1.x()) and (p.x() > p2.x()):
            return False
        if (p.x() < p1.x()) and (p.x() < p2.x()):
            return False
        if (p.y() > p1.y()) and (p.y() > p2.y()):
            return False
        if (p.y() < p1.y()) and (p.y() < p2.y()):
            return False
        return True
